irohad/ametsuchi/impl/postgres_block_index.cpp

```diff
@@ -0,0 +1,117 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ametsuchi/impl/postgres_block_index.hpp"
+
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/adaptor/indexed.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+#include <boost/range/algorithm/for_each.hpp>
+#include <boost/range/numeric.hpp>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "cryptography/ed25519_sha3_impl/internal/sha3_hash.hpp"
+#include "model/commands/transfer_asset.hpp"
+#include "model/sha3_hash.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+
+    PostgresBlockIndex::PostgresBlockIndex(pqxx::nontransaction &transaction)
+        : transaction_(transaction),
+          log_(logger::log("PostgresBlockIndex")),
+          execute_{makeExecute(transaction_, log_)} {}
+
+    void PostgresBlockIndex::index(const model::Block &block) {
+      const auto &height = std::to_string(block.height);
+      boost::for_each(block.transactions | boost::adaptors::indexed(0),
+                      [&](const auto &tx) {
+                        const auto &creator_id = tx.value().creator_account_id;
+                        const auto &hash = iroha::hash(tx.value()).to_string();
+                        const auto &index = std::to_string(tx.index());
+
+                        // tx hash -> block where hash is stored
+                        execute_("INSERT INTO height_by_hash(hash, height) VALUES ("
+                                 + transaction_.quote(
+                                pqxx::binarystring(hash.data(), hash.size()))
+                                 + ", " + transaction_.quote(height) + ");");
+
+                        this->indexAccountIdHeight(creator_id, height);
+
+                        // to make index account_id:height -> list of tx indexes
+                        // (where tx is placed in the block)
+                        execute_(
+                            "INSERT INTO index_by_creator_height(creator_id, "
+                            "height, index) "
+                            "VALUES ("
+                            + transaction_.quote(creator_id) + ", "
+                            + transaction_.quote(height) + ", "
+                            + transaction_.quote(index) + ");");
+
+                        this->indexAccountAssets(
+                            creator_id, height, index, tx.value().commands);
+                      });
+    }
+
+    void PostgresBlockIndex::indexAccountIdHeight(const std::string &account_id,
+                                                  const std::string &height) {
+      execute_(
+          "INSERT INTO height_by_account_set(account_id, height) "
+          "VALUES ("
+          + transaction_.quote(account_id) + ", " + transaction_.quote(height)
+          + ");");
+    }
+
+    void PostgresBlockIndex::indexAccountAssets(
```

Comment:
If some problems will occasion method has to send it via return value.

Reply:
Do we need it here right now? These methods are only used in 'index', but it has no return value, so it is not possible to pass an error outside BlockIndex